src/librbd/operation/EnableFeaturesRequest.cc

```diff
@@ -0,0 +1,435 @@
+               << dendl;
+    return handle_finish(*result);
+  }
+
+  send_set_features();
+  return nullptr;
+}
+
+template <typename I>
+void EnableFeaturesRequest<I>::send_set_features() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << ": new_features="
+		 << m_new_features << ", features_mask=" << m_features_mask
+		 << dendl;
+
+  librados::ObjectWriteOperation op;
+  librbd::cls_client::set_features(&op, m_new_features, m_features_mask);
+
+  using klass = EnableFeaturesRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_set_features>(this);
+  int r = image_ctx.md_ctx.aio_operate(image_ctx.header_oid, comp, &op);
+  assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+Context *EnableFeaturesRequest<I>::handle_set_features(int *result) {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(cct) << "failed to update features: " << cpp_strerror(*result)
+               << dendl;
+    return handle_finish(*result);
+  }
+
+  send_create_object_map();
+  return nullptr;
+}
+
+template <typename I>
+void EnableFeaturesRequest<I>::send_create_object_map() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  if (((image_ctx.features & RBD_FEATURE_OBJECT_MAP) != 0) ||
+      ((m_features & RBD_FEATURE_OBJECT_MAP) == 0)) {
+    send_enable_mirror_image();
+    return;
+  }
+
+  Context *ctx = create_context_callback<
+    EnableFeaturesRequest<I>,
+    &EnableFeaturesRequest<I>::handle_create_object_map>(this);
+
+  image::CreateObjectMapRequest<I> *req =
+    image::CreateObjectMapRequest<I>::create(&image_ctx, ctx);
+  req->send();
+}
+
+template <typename I>
+Context *EnableFeaturesRequest<I>::handle_create_object_map(int *result) {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(cct) << "failed to create object map: " << cpp_strerror(*result)
+               << dendl;
+    return handle_finish(*result);
+  }
+
+  send_enable_mirror_image();
+  return nullptr;
+}
+
+template <typename I>
+void EnableFeaturesRequest<I>::send_enable_mirror_image() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  if (!m_enable_mirroring) {
+    send_append_op_event();
+    return;
+  }
+
+  Context *ctx = create_context_callback<
+    EnableFeaturesRequest<I>,
+    &EnableFeaturesRequest<I>::handle_enable_mirror_image>(this);
+
+  image::MirrorEnableRequest<I> *req =
+    image::MirrorEnableRequest<I>::create(&image_ctx.md_ctx, image_ctx.id, ctx);
```

Comment:
Interesting question :-) I do this just because it was in the initial (sync) implementation.

https://github.com/ceph/ceph/pull/8204/commits/ba6c0c9d4cc7675a71741bddf2e8266117b689c8

I thought there were some reasons to do this...